src/javaharness/java/arcs/api/PECInnerPortImpl.java

```diff
@@ -121,9 +152,21 @@ public void SynchronizeProxy(StorageProxy storageProxy, Consumer<PortableJson> c
         postMessage(message);
     }
 
+    @Override
+    public void Render(NativeParticle particle, String slotName, PortableJson content) {
+        PortableJson message = jsonParser.parse("{}");
+        message.put(MESSAGE_TYPE_FIELD, RENDER_MSG);
+        PortableJson body = jsonParser.parse("{}");
+        body.put(PARTICLE_FIELD, mapper.identifierForThing(new Thing<NativeParticle>(particle)));
+        body.put(SLOT_NAME_FIELD, slotName);
+        body.put(CONTENT_FIELD, content);
+        message.put(MESSAGE_BODY_FIELD, body);
+        postMessage(message);
+    }
+
     private void postMessage(PortableJson message) {
-        // TODO: add support for PEC messages in pipes-shell-2 bus.
-        // message.put(MESSAGE_PEC_ID_FIELD, this.id);
-        // shellApi.postMessage(jsonParser.stringify(message));
+        environment.sendMessageToArcs(
+                    "{\"message\": \"pec\", \"id\": \"" + this.id +
```

Comment:
Perhaps we should use an Arcs Shell Data Bus message builder class, potentially one we write ourselves, to help generate this sort of formatted thing down the road, I think we do this in various places outside of the Java PEC as well, for work @yuangu-google has done.

Reply:
for now i just switched to use Json object instead of string concatenation, but i'll talk to Yuan about that. thanks!

Reply:
Hi Maria,

I am using this package to construct Jsons. https://stleary.github.io/JSON-java/org/json/JSONObject.html which works fine for me.

Reply:
Java harness is using: https://github.com/PolymerLabs/arcs/blob/master/src/javaharness/java/arcs/api/PortableJson.java and https://github.com/PolymerLabs/arcs/blob/master/src/javaharness/java/arcs/webimpl/PortableJsonJsImpl.java
we should wait for Ray to get back, if we want to consider an alternative.

But iiuc Walter was suggesting we implement an arcs-shell-bus message builder on top of json, which we probably should do no matter what json we use?

Reply:
Hmm, at this moment I am not sure what fields would be used in the message. So maybe we can wait until we have more insights on the shared part of the messages contents.